### PR TITLE
Retrieve function logs from controller

### PR DESCRIPTION
**Pull request by life1347**
This PR tries to let controller establishes a reverse proxy to log database, and redirect query command send from client to database and then proxy back the db responses. So that we dont need to expose database to the public network.

### CODE REVIEW

File: controller/functionApi.go [hunk 1]
```diff
@@ -171,3 +175,35 @@ func (api *API) FunctionApiDelete(w http.ResponseWriter, r *http.Request) {
...
+	director := func(req *http.Request) {
+		// only replace url Scheme and Host to remote influxDB
+		// and leave query string intact
+		req.URL.Scheme = svcUrl.Scheme
+		req.URL.Host = svcUrl.Host
+		req.URL.Path = svcUrl.Path
+		// set up http basic auth for database authentication
+		req.SetBasicAuth(dbCnf.username, dbCnf.password)
+	}
+	rrt := router.NewRetryingRoundTripper(10, 50*time.Millisecond)
```
Comment (soamvasani) [inline controller/functionApi.go]: Why do we need this? When does the default roundtripper not work?

File: INSTALL.md [hunk 2]
```diff
@@ -168,6 +151,18 @@ particular pod:
   $ fission function logs --name hello --pod <pod name>
 ```
 
+* (Optional) Edit `fission-logger.yaml` to add Influxdb username and password and following env to controller deployment.  Then apply `fission-logger.yaml` and new controller deployment.
```
Comment (soamvasani) [inline INSTALL.md]: I suggest just removing this section. We should generate a random password in the helm chart (the plan is to move to that as the main recommended installation method).

File: controller/api.go [hunk 3]
```diff
@@ -78,6 +105,7 @@ func (api *API) Serve(port int) {
 	r.HandleFunc("/v1/functions/{function}", api.FunctionApiGet).Methods("GET")
 	r.HandleFunc("/v1/functions/{function}", api.FunctionApiUpdate).Methods("PUT")
 	r.HandleFunc("/v1/functions/{function}", api.FunctionApiDelete).Methods("DELETE")
+	r.HandleFunc("/v1/functions/{function}/logs/{dbType}", api.FunctionLogsApiPost).Methods("POST")
```
Comment (soamvasani) [inline controller/api.go]: Thinking about this path: It seems the `{function}` variable is not really being used? The query from the CLI has the function name. 

Also this isn't really part of the versioned API. The API is still influxdb's, this is just a proxy.

How about simply `/proxy/{dbType}`?

File: controller/functionApi.go (hunk 1 above)
Comment (life1347) [inline controller/functionApi.go]: In fact, just want to follow the proxy setting in the router, not really need it. I'll remove it.

**Comment by life1347** [comment]
The test process:

1. Follow the step described in INSTALL.md to set up persistent Function Logs
2. Compile and deploy the new version of controller
3. Compile fission CLI
4. Create a function and route for function`hello1`
5. Try to curl the function
6. Once the curl request is success, try with the following commands to retrieve logs from controller
```
fission fn logs --name hello1 -f -d
fission fn pods --name hello1
```

File: fission/logdb/influxdb.go [hunk 4]
```diff
@@ -17,47 +17,35 @@ limitations under the License.
...
-	if err != nil {
-		log.Fatal(err)
-		return InfluxDB{}, err
-	}
-
-	return InfluxDB{
-		dbClient: dbClient,
-	}, nil
+func NewInfluxDB(serverURL string) (InfluxDB, error) {
+	return InfluxDB{endpoint: serverURL}, nil
```
Comment (soamvasani) [inline fission/logdb/influxdb.go]: Can't we use the influxdb client, but specify the new URL in the HTTPConfig Addr field?  That way we won't have to implement the query() function too.

(Looking at the code https://github.com/influxdata/influxdb/blob/master/client/v2/client.go#L82, it looks like it should just work.)

File: fission/logdb/influxdb.go [hunk 5]
```diff
@@ -17,47 +17,35 @@ limitations under the License.
...
 }
 
 type InfluxDB struct {
-	dbClient influxdbClient.Client
+	endpoint string
 }
 
 func (influx InfluxDB) GetPods(filter LogFilter) ([]string, error) {
 	query := fmt.Sprintf("select * from \"log\" where \"funcuid\" = '%s' group by \"pod\"", filter.FuncUid)
```
Comment (soamvasani) [inline fission/logdb/influxdb.go]: Can we remove this Sprintf, which is vulnerable to "sql" injection, and instead use `influxdb.NewQueryWithParameters`?

File: fission/logdb/influxdb.go (hunk 4 above)
Comment (life1347) [inline fission/logdb/influxdb.go]: Influxdb client will replace the Addr url path before sending the query string to remote host.(https://github.com/influxdata/influxdb/blob/master/client/v2/client.go#L475). So we need to write our own read-only db client. (In fact, this issue took some time to debug when I tried to use proxy server solution. lol)

File: fission/logdb/influxdb.go (hunk 5 above)
Comment (life1347) [inline fission/logdb/influxdb.go]: oh, you are right! I'll take a look at it.

**Comment by soamvasani** [comment]
Thanks, @life1347 ! The final changes look great.

**Pull request merged**